internal/ip/ip.go

```diff
@@ -0,0 +1,179 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ip
+
+import (
+	"fmt"
+	"net"
+	"syscall"
+
+	"github.com/pkg/errors"
+	d "github.com/uber/arachne/defines"
+	"github.com/uber/arachne/internal/log"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/google/gopacket/pcap"
+	"go.uber.org/zap"
+)
+
+// Conn represents the underyling functionality to send and recv Arachne echo requests.
+type Conn struct {
+	SrcAddr net.IP
+	AF      int
+	sendFD  int
+	recvSrc gopacket.PacketDataSource
+}
+
+// Close is used to close a Conn's send file descriptor and Recv packet source.
+func (c *Conn) Close() error {
+	return syscall.Close(c.sendFD)
+}
+
+// NextPacket gets the next available packet from the PacketDataSource.
+func (c *Conn) NextPacket() (gopacket.Packet, error) {
+	data, _, err := c.recvSrc.ReadPacketData()
+	if err != nil {
+		return nil, err
+	}
+	return gopacket.NewPacket(data, layers.LinkTypeEthernet, gopacket.DecodeOptions{Lazy: true}), nil
+}
+
+func ipToSockaddr(family int, ip net.IP, port int) (syscall.Sockaddr, error) {
+	switch family {
+	case syscall.AF_INET:
+		if len(ip) == 0 {
+			ip = net.IPv4zero
+		}
+		ip4 := ip.To4()
+		if ip4 == nil {
+			return nil, &net.AddrError{Err: "non-IPv4 address", Addr: ip.String()}
+		}
+		sa := &syscall.SockaddrInet4{Port: port}
+		copy(sa.Addr[:], ip4)
+		return sa, nil
+	case syscall.AF_INET6:
+		if len(ip) == 0 || ip.Equal(net.IPv4zero) {
+			ip = net.IPv6zero
+		}
+		ip6 := ip.To16()
+		if ip6 == nil {
+			return nil, &net.AddrError{Err: "non-IPv6 address", Addr: ip.String()}
+		}
+		sa := &syscall.SockaddrInet6{Port: port}
+		copy(sa.Addr[:], ip6)
+		return sa, nil
+	}
+	return nil, &net.AddrError{Err: "invalid address family", Addr: ip.String()}
+}
+
+// Sendto operates on a Conn file descriptor and mirrors the Sendto syscall.
+func (c *Conn) Sendto(b []byte, to net.IP) error {
+	sockAddr, err := ipToSockaddr(c.AF, to, 0)
+	if err != nil {
+		return err
+	}
+
+	return syscall.Sendto(c.sendFD, b, 0, sockAddr)
+}
+
+// getSendSocket will create a raw socket for sending data.
+func getSendSocket(af int) (int, error) {
+	fd, err := syscall.Socket(af, syscall.SOCK_RAW, syscall.IPPROTO_RAW)
+	if err != nil {
+		return 0, err
+	}
+
+	if err = syscall.SetsockoptInt(fd, syscall.IPPROTO_IP, syscall.IP_HDRINCL, 1); err != nil {
+		return 0, err
+	}
+
+	return fd, nil
+}
+
+func getRecvSource(listenPort uint32, intf string) (gopacket.PacketDataSource, error) {
+	// Filter for tcp packets coming into ListenPort, http, or HTTPS and contain a SYN flag
+	filter := fmt.Sprintf("(dst port %d or src port %d or src port %d) and (tcp[13] & 2 != 0)",
```

Comment:
This is checking that the SYN flag is set, regardless of any other flags. So, it matches any SYN, or SYN+ACK packets (which are the only ones arachne needs to receive). Filtering with BPF is much more efficient that in Arachne's Go code, so I wanted to be as specific as possible here.